flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/PartitionedFileWriter.java

```diff
@@ -0,0 +1,211 @@
+	private int numRegions;
+
+	/** Current subpartition to write. Buffer writing must be in subpartition order within each region. */
+	private int currentSubpartition;
+
+	/** Whether all index data can be cached in memory or not. */
+	private boolean canCacheAllIndexData = true;
+
+	/** Whether this file writer is finished. */
+	private boolean isFinished;
+
+	public PartitionedFileWriter(String basePath, int numSubpartitions) {
+		checkArgument(basePath != null, "Base path must not be null.");
+		checkArgument(numSubpartitions > 0, "Illegal number of subpartitions.");
+
+		this.numSubpartitions = numSubpartitions;
+		this.subpartitionBytes = new long[numSubpartitions];
+		this.subpartitionBuffers = new int[numSubpartitions];
+		this.dataFilePath = new File(basePath + PartitionedFile.DATA_FILE_SUFFIX).toPath();
+		this.indexFilePath = new File(basePath + PartitionedFile.INDEX_FILE_SUFFIX).toPath();
+
+		this.indexBuffer = ByteBuffer.allocate(100 * 1024 * PartitionedFile.INDEX_ENTRY_SIZE);
+		indexBuffer.order(PartitionedFile.DEFAULT_BYTE_ORDER);
+	}
+
+	/**
+	 * Opens the {@link PartitionedFile} for writing.
+	 *
+	 * <p>Note: The caller is responsible for releasing the failed {@link PartitionedFile} if any exception
+	 * occurs.
+	 */
+	public void open() throws IOException {
+		checkState(dataFileChannel == null && indexFileChannel == null, "Partitioned file is already opened.");
+
+		dataFileChannel = openFileChannel(dataFilePath);
+		indexFileChannel = openFileChannel(indexFilePath);
+	}
+
+	private FileChannel openFileChannel(Path path) throws IOException {
+		return FileChannel.open(path, StandardOpenOption.CREATE_NEW, StandardOpenOption.WRITE);
+	}
+
+	/**
+	 * Persists the region index of the current data region and starts a new region to write.
+	 *
+	 * <p>Note: The caller is responsible for releasing the failed {@link PartitionedFile} if any exception
+	 * occurs.
+	 */
+	public void startNewRegion() throws IOException {
+		checkState(!isFinished, "File writer is already finished.");
+		checkState(dataFileChannel != null && indexFileChannel != null, "Must open the partitioned file first.");
+
+		writeRegionIndex();
+	}
+
+	private void writeRegionIndex() throws IOException {
+		if (Arrays.stream(subpartitionBytes).sum() > 0) {
+			for (int subpartition = 0; subpartition < numSubpartitions; ++subpartition) {
+				if (!indexBuffer.hasRemaining()) {
+					flushIndexBuffer();
+					indexBuffer.clear();
+					canCacheAllIndexData = false;
+				}
+
+				indexBuffer.putLong(totalBytesWritten);
+				indexBuffer.putInt(subpartitionBuffers[subpartition]);
+				totalBytesWritten += subpartitionBytes[subpartition];
+			}
+
+			++numRegions;
+			currentSubpartition = 0;
+			Arrays.fill(subpartitionBytes, 0);
+			Arrays.fill(subpartitionBuffers, 0);
+		}
+	}
+
+	private void flushIndexBuffer() throws IOException {
+		if (indexBuffer.position() > 0) {
+			indexBuffer.flip();
+			indexFileChannel.write(indexBuffer);
```

Comment:
I think here we still need to iterates till the whole buffer is write (writeFully). The dataFileChannel has also done this.